keps/sig-api-machinery/20180731-crd-pruning.md

```diff
@@ -0,0 +1,610 @@
+  // Defaults to false. 
+  // Setting this field to true is considered an alpha API.
+  // Note: this will default to true in version v1.
+  PruneUnknownFields bool
+}
+```
+
+I.e. for `apiextensions.k8s.io/v1beta1` this will default to `false` for backwards compatibility.
+
+For `apiextensions.k8s.io/v1` we will change the default to `true` and forbid `false` during creation and updates if it has been `true` before. In `v1` the only way to opt-out from pruning is via setting `x-kubernetes-preserve-unknown-fields: true` in the schema. An empty or unspecified schema will prune everything.
+
+### References
+
+* Old pruning implementation PR https://github.com/kubernetes/kubernetes/pull/64558, to be adapted 
+* [OpenAPI v3 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md)
+* [JSON Schema](http://json-schema.org/)
+
+### Test Plan
+
+**blockers for alpha:**
+
+We default `PruneUnknownFields` to false and hence switch off the whole code path doing pruning. This reduces risk for everybody not using this alpha feature.
+
+* we add unit tests for the general pruning algorithm
+* we add apiextensions-apiserver integration tests to
+  * verify that the pruning feature is actually off if `PruneUnknownFields` is false.
+  * verify that `PruneUnknownFields`  is defaulted to false.
+  * verify that pruning happens if `PruneUnknownFields` is true, for all versions in the CRD according to the schema of the respective version.
+  * verify that `metadata`, `apiVersion`, `kind` are preserved if `PruneUnknownFields` is true and there is no schema given in the CRD.
+  
+**blockers for beta:**
+
+* we implement and verify that `x-kubernetes-embedded-resource` and `x-kubernetes-preserve-unknown-fields` work as expected.
+* we add apiextensions-apiserver integration tests to
+  * verify that pruning happens on incoming request payloads, on read from storage and after calling mutating admission webhooks.
+
+**blockers for GA:**
+
+* we verified that performance of pruning is adequate and not considerably reducing throughput.
+
+### Graduation Criteria
+
+* the test plan is fully implemented for the respective quality level
+
+### Upgrade / Downgrade Strategy
+
+* setting `PruneUnknownFields` to true is considered alpha quality in 1.15.
+* downgrading to 1.14 will lose `pruneUnknownFields: true`, but that's acceptable.
+* downgrading from 1.16 (where pruning might be beta) to 1.15 will keep the same behaviour as we don't feature gate `pruneUnknownFields: true`.
+* upgrading from 1.14 will default to `pruneUnknownFields: false` and hence change no behaviour.
+* upgrading from 1.15 will keep the value and hence change no behaviour.
+* when `v1` of `apiextensions.k8s.io` is added, we will keep the old pruning behaviour for CRDs created in `v1beta1` with `pruneUnknownFields: false`, but enforce `pruneUnknownFields: true` for every newly create `v1` CRD. Hence, we keep backwards compatibility.
```

Comment:
> enforce `pruneUnknownFields: true` for every newly create `v1` CRD

I know we want to default to pruning in v1, but does this mean we don't allow `pruneUnknownFields: false` at all in new v1 CRDs? If so, explaining the rationale for forcing use of the openapi extension attribute would be helpful

Reply:
We allow `x-kubernetes-preserve-unknown-fields: true` at the root level such that only metadata is pruned. So we only switch opt-in and opt-out. `pruneUnknownFields` is just the verhicle to do the transition.

Will add something along these lines.

Reply:
done